Best practices SPD-306 (#45)
sara-nl/spiderdocs: source/Pages/best_practices.rst

.. _best-practices:
*****************
Best practices
*****************

.. Tip:: This is a best practices section to help you achieve maximum performance for your jobs on Spider. In this page you will learn which options are most efficient for:

     * your software installation on Spider
     * storing your data on the platform
     * managing your data on Spider  
     * running a large amount of jobs on Spider

==========     
Background
==========

Spider is continuously expanding as a unique data processing and collaboration platform. The growing demand for storage space, the diversity in applications and the various ways the system is used, all bring some technical challenges. For example, if you're doing a lot of IO-operations (reading and writing files) in your workflows, you should be mindful where these operations are performed as some options can significantly affect the performance of your jobs and the system load. 

> Our local file system on Spider is CephFS and is suitable as a staging area for your data before or after analysing it. CephFS hosts both your home and project directories on Spider. It is designed for efficient IO of large files, but when dealing with many small files the file system performance can be degraded. This is because CephFS relies on a parallel distributed system that involves many disks to store the data itself and metadata servers to store the files metadata. As a result, when you operate on many small files or run code from python environments, the system response can become slow for you and other users on the platform. 

Author: Good point. It is rephrased as python environments to be an example of small file repos.

In order to improve the performance of scientific workflows on Spider's storage, we have prepared this best practices guide that includes several tips to install, store and analyse your data efficiently.

.. _software-practices:

Software installation practices
===============================

If your software loads a large number of small files upon execution in your jobs then you may see poor I/O performance even if the total software size is not that big. There are ways to mitigate potential bottlenecks, such as the use of CVMFS or containers technology.

Here is an overview of the features and capabilities of some of the software installation options supported on Spider followed by some general guidelines:

==============================================   ==============   ==============   =============   ======================

> Feature                                          CephFS           Apptainer        Lumi            Softdrive

Reviewer: Do you mean LUMI container wrapper? I think here should refer to the tool, instead of LUMI.

Author: Good point. I'll fix it

We also advice you to use the scratch file systems as fast temporary storage that can be used while running a job. Each of the Spider worker nodes has a large scratch area on local SSD. Any data that you wish to keep should be written to other storage backends such as dCache. The scratch areas are ideal for retrieving the input of a job from dCache during execution or for applications that generate lots of intermediate files that are consumed by other parts of the processing or for generating the job output before copying it back to dCache. More about scratch can be found in our section :ref:`How to use the temporary disk space <scratch-fs>`.

> In cases that you have multiple jobs that need to access a single set of files that is too large to copy over to scratch, it is possible to use CephFS on home or project space locations for temporary storing your data, but take into account its limitations such as slow throughput and short capacity compared to dCache.

Reviewer: Shall we say "but take into account its limitations compared to dCache in terms of throughput and capacity"?  Phrases like "slow throughput and short capacity might give a negative impression of CephFS.

Author: Good point. Fixed.

For advanced users, who are familiar with mpi operations, we also offer an a MPI-based tool for managing datasets such as copying files across the different home and project space folders on the local file system. The MPI-based tool is much faster and efficient than the common `cp` operations. Example usage for parallel copying of files using this method can be found in the :ref:`mpifileutils <mpi-file-utils>` section.

.. _running-many-jobs:

Running a large amount of jobs 
==============================

> When running a large amount of jobs it can be difficult to keep track of the state of these jobs or resume failed tasks that were prematurely canceled due to time limit. Another challenge when designing high-throughput workflows that execute a specific application for many different parameter combinations, is reducing the large scheduling overhead and waiting times in the queue. 

Reviewer 2: ```suggestion
High-throughput workflows that execute a specific application for many different parameter combinations, often requires the submission of many jobs. When running a large amount of jobs it can be difficult to keep track of the status of these jobs or resume failed jobs that were prematurely canceled (e.g. due to time limit). Another challenge is reducing the large scheduling overhead and waiting times in the queue. 
```

An overview of the features and capabilities of some of the options for running a large amount of jobs on Spider is presented below.

==============================================   ==============   ==============   =============
Feature                                          Array jobs       Picas            Snakemake
==============================================   ==============   ==============   =============  
High speed & Low load on the system              No               Yes              Somewhat  
Scales to hundreds, thousands of jobs and more   No               Yes              Somewhat

> Available outside Spider                         No               Yes              No

Reviewer: Correct me if I am wrong, isn't Snakemake also available elsewhere?

Author: Good point. It is rephrased as 'transcends spider' to explain whether it is available in single slurm cluster or across many.

> An overview of the features and capabilities of some of the options for running a large amount of jobs on Spider is presented below.
> 
> ==============================================   ==============   ==============   =============
> Feature                                          Array jobs       Picas            Snakemake
> ==============================================   ==============   ==============   =============  
> High speed & Low load on the system              No               Yes              Somewhat  
> Scales to hundreds, thousands of jobs and more   No               Yes              Somewhat
> Available outside Spider                         No               Yes              No
> Easy setup                                       Yes              Somewhat         Somewhat    
> Handles easily dependencies between tasks        No               Somewhat         Yes
> Error recovery                                   No               Yes              No
> ==============================================   ==============   ==============   =============

Reviewer 2: ```suggestion
An overview of the features and suitability of some of the options for running a large amount of jobs on Spider is presented below.
==============================================   ==============================================================   ==============================   ======================================
Feature                                          `Slurm job arrays <https://slurm.schedmd.com/job_array.html>`_   :ref:`PiCaS <picas-on-spider>`   :ref:`Snakemake <snakemake-on-spider>`
==============================================   ==============================================================   ==============================   ======================================  
High speed & low load on the system              No                                                               Yes                              Moderate
Scales to hundreds, thousands of jobs and more   No                                                               Yes                              Moderate
Available outside Spider                         No                                                               Yes                              No
Easy setup                                       Yes                                                              Moderate                         Moderate    
Handles easily dependencies between tasks        No                                                               Moderate                         Yes
Error recovery                                   No                                                               Yes                              No
==============================================   ==============================================================   ==============================   ======================================
```

Which practice for running a large amount of jobs should I use?
---------------------------------------------------------------

> The first option to check when running a large amount of jobs is whether the software you're using comes with a built-in option for managing your workloads on a Slurm-based cluster. Alternatively, an easy way to submit several independent jobs with one command is the use of Array jobs. Array jobs, however do not scale well for more than a few hudrends of jobs. In this case, you can use external tools for managing your workloads, such as PiCaS or Snakemake.

Reviewer 2: ```suggestion
The first option to check when running a large amount of jobs is whether the software you're using comes with a built-in option for managing your workloads on a Slurm-based cluster. Alternatively, an easy way to submit several independent jobs with one command is the use of Job arrays. Job arrays, however do not scale well for more than a few hundreds of jobs. In this case, you can use external tools for managing your workloads, such as PiCaS or Snakemake.
```

> PiCaS works as a queue, providing a mechanism to step through the work one task at a time. It is also a pilot job system, indicating that the client communicates with the PiCaS server to fetch work, instead of having that work specified in a job (or similar) file.  As every application needs different parameters, PiCaS has a flexible data structure that allows users to save different types of data. PiCaS can handle thousands or millions of tasks, it has an easy query mechanism to search among your tasks and is accessible from any platform via a Restful HTTP API. Here you can find instructions for using :ref:`PiCaS <picas-on-spider>`.

Reviewer 2: ```suggestion
PiCaS works as a queue, providing a mechanism to step through the work one task at a time. It is also a `pilot job <https://doc.grid.surfsara.nl/en/latest/Pages/Practices/pilot_jobs.html>`_  system, indicating that the client communicates with the PiCaS server to fetch work, instead of having that work specified in a job (or similar) file.  As every application needs different parameters, PiCaS has a flexible data structure that allows users to save different types of data. PiCaS can handle thousands or millions of tasks, it has an easy query mechanism to search among your tasks and is accessible from any platform via a Restful HTTP API. Here you can find instructions for using :ref:`PiCaS <picas-on-spider>`.
```

> When your application involves several steps connected in a workflow that each need to be submitted as independent tasks, Snakemake is tool for defining, managing and executing workflows with multiple steps and complex dependencies. There are possibilities to combine PiCaS and Snakemake to enable workflow automation and run many jobs and subtasks efficiently and fast. Please contact our :ref:`our helpdesk <helpdesk>` if you need help with automating your workloads on Spider.

Reviewer 2: ```suggestion
When your application involves several steps connected in a workflow that each need to be submitted as independent tasks, you may consider using :ref:`Snakemake <snakemake-on-spider>`. Snakemake is a python-based workflow managment tool for defining, managing and executing workflows with multiple steps and complex dependencies. There are possibilities to combine PiCaS and Snakemake to enable workflow automation and run many jobs and subtasks efficiently and fast. Please contact our :ref:`our helpdesk <helpdesk>` if you need help with automating your workloads on Spider.
```
